### PR TITLE
refactor(expr): use a wrapper for `eval_checked`

### PR DESCRIPTION
**Pull request by BugenZhao**
Signed-off-by: Bugen Zhao <i@bugenzhao.com>I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Adopt the wrapper pattern from executors to expressions, implement checked evaluation (`eval_checked`) with a `CheckedExpression` wrapper, instead of a method with default implementations, which may be forgotten to call sometimes.

Furthermore, we can use this pattern to resolve https://github.com/risingwavelabs/risingwave/issues/11586.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/expr/src/expr/build.rs [hunk 1]
```diff
@@ -70,6 +71,15 @@ pub fn build_from_prost(prost: &ExprNode) -> Result<BoxedExpression> {
     }
 }
 
+/// Build an expression from protobuf with wrappers.
+pub fn build_from_prost(prost: &ExprNode) -> Result<BoxedExpression> {
+    let expr = build_from_prost_inner(prost)?;
+
+    let checked = CheckedExpression(expr);
+
+    Ok(checked.boxed())
```
Comment (stdrc) [inline src/expr/src/expr/build.rs:80]: Why insert empty line between every lines😁

**Review by stdrc — APPROVED** [review]
Rest LGTM

**Review by wangrunji0408 — APPROVED** [review]
LGTM!

In fact, I think the assertions are no longer needed for most macro-generated expressions, since we are quite confident in the correctness of the template. Or we can move the assertion into the template to avoid additional boxing overhead. 🤔

File: src/expr/src/expr/build.rs (hunk 1 above)
Comment (BugenZhao) [reply src/expr/src/expr/build.rs]: Will soon be refactored in the next PR. 🤣

**Comment by BugenZhao** [comment]
> I think the assertions are no longer needed for most macro-generated expressions

I agree. 🫡 This PR is to introduce the wrapper pattern more than refactoring `checked` itself, which I will utilize to resolve #11586.

**Pull request merged**